include/deal.II/base/tensor.h

```diff
@@ -1739,6 +1783,57 @@ DEAL_II_CUDA_HOST_DEV constexpr DEAL_II_ALWAYS_INLINE
 }
 
 
+namespace internal
+{
+  namespace TensorImplementation
+  {
+    template <int rank,
+              int dim,
+              typename Number,
+              typename OtherNumber,
+              typename std::enable_if<
+                !std::is_integral<
+                  typename ProductType<Number, OtherNumber>::type>::value,
+                int>::type = 0>
+    DEAL_II_CONSTEXPR DEAL_II_CUDA_HOST_DEV inline DEAL_II_ALWAYS_INLINE
+                      Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type>
+                      division_operator(const Tensor<rank, dim, Number> &t,
+                                        const OtherNumber &              factor)
+    {
+      using common_type = typename ProductType<Number, OtherNumber>::type;
+      Tensor<rank, dim, common_type> tt;
+      const auto                     inverse_factor =
+        static_cast<common_type>(1.) / static_cast<common_type>(factor);
+      // recurse over the base objects
+      for (unsigned int d = 0; d < dim; ++d)
+        tt[d] = t[d] * inverse_factor;
+      return tt;
+    }
+
+
+    template <int rank,
+              int dim,
+              typename Number,
+              typename OtherNumber,
+              typename std::enable_if<
+                std::is_integral<
+                  typename ProductType<Number, OtherNumber>::type>::value,
+                int>::type = 0>
+    DEAL_II_CONSTEXPR DEAL_II_CUDA_HOST_DEV inline DEAL_II_ALWAYS_INLINE
+                      Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type>
+                      division_operator(const Tensor<rank, dim, Number> &t,
+                                        const OtherNumber &              factor)
+    {
+      Tensor<rank, dim, typename ProductType<Number, OtherNumber>::type> tt;
+      // recurse over the base objects
+      for (unsigned int d = 0; d < dim; ++d)
+        tt[d] = t[d] / factor;
+      return tt;
+    }
+  } // namespace TensorImplementation
+} // namespace internal
```

Comment:
Can't you express the member function `/=` in terms of this function (or the other way around) to only have one implementation class? For example, you could pass in `*this` as the operand in this function? This puts some duty on the compiler to properly handle return types but I think it will be fine.

Reply:
I could unify the two by having a 
```
division_operator(const Tensor &input, const Number scalar, Tensor output)
{
  for (i ...)
    output[i] = input[i]/scalar;
}
```
interface and
```
operator/(const Tensor &t, const Number scalar)
{
  Tensor tmp;
  division_operator(t, scalar, tmp);
 return tmp;
}

Tensor::operator/=(const Number scalar)
{
  division_operator(t, scalar, t);
}
```

I am just not sure if we copy unnecessarily for `operator/=` then.